[PropertyAccess] Show property path in all exception messages

mpajunen opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | #12987 |
| License | MIT |
| Doc PR | N/A |

Add the property path to the exception message when trying to access a property of a scalar variable or null. For example:

``` php
$accessor->getValue((object) ['foo' => null], 'foo.bar'); // Throws UnexpectedTypeException
$old = 'Expected argument of type "object or array", "NULL" given'; // Old message
$new = 'PropertyAccessor requires a graph of objects or arrays to operate on, but it found type "NULL" while trying to traverse path "foo.bar" at property "bar".'; // New message
```

Checking the exception messages was added to the existing tests.

~~To make it possible to customize the exception messages, constructor was removed from `UnexpectedTypeException`. Instead the exception messages are built where the exception is thrown. (None of the other exception classes of the component had custom constructors.)~~

~~There are no BC breaks -- unless removing the constructor of an exception class is considered one.~~

The original PR included a potential BC break but the structure was changed somewhat.

The current version deprecates the old constructor for `UnexpectedTypeException`, but a BC layer is included.

fabpot commented:
:+1:

Tobion commented on src/Symfony/Component/PropertyAccess/Exception/UnexpectedTypeException.php:
```diff
@@ -18,8 +18,4 @@
  */
 class UnexpectedTypeException extends RuntimeException
 {
-    public function __construct($value, $expectedType)
```
this is a bc break

mpajunen replied:
Yep, it is a theoretical BC break, hence my note at the bottom of the original PR comment. I don't see a straightforward way around it:

- Using a new subclass won't work because the constructor is already overloaded.
- Using another exception class entirely would be a worse BC break.
- Adding optional constructor arguments might be a smaller BC break, but those arguments would make no sense when the exception is used in `PropertyPath`.

I can of course change this somehow, or add the BC break notes, or scrap the PR entirely.

Tobion replied:
I'd suggest to
1. change the exception in PropertyPath to `Symfony\Component\PropertyAccess\Exception\InvalidArgumentException` because that is more consistent with https://github.com/mpajunen/symfony/blob/property-access-exception-messages/src/Symfony/Component/PropertyAccess/PropertyAccessor.php#L57 and also makes more sense since it is a RuntimeException isntead of a LogicException (and should not be considered BC break since it's not intended to be catched).
2. adjust UnexpectedTypeException  with optional extra params

mpajunen replied:
That would work, I'll make the changes. Good point that the `PropertyPath` exception is not meant to be caught anyway.

Tobion commented on src/Symfony/Component/PropertyAccess/PropertyAccessor.php:
```diff
@@ -97,7 +97,12 @@ public function setValue(&$objectOrArray, $propertyPath, $value)
-                    throw new UnexpectedTypeException($objectOrArray, 'object or array');
+                    throw new UnexpectedTypeException(sprintf(
+                        'Expected object or array, "%s" found when trying to write property "%s" of path "%s".',
+                        is_object($objectOrArray) ? get_class($objectOrArray) : gettype($objectOrArray),
```
this condition makes no sense, since we already know that `$objectOrArray` is NOT an object

Tobion commented on src/Symfony/Component/PropertyAccess/PropertyAccessor.php:
```diff
@@ -221,7 +226,12 @@ private function &readPropertiesUntil(&$objectOrArray, PropertyPathInterface $pr
-                throw new UnexpectedTypeException($objectOrArray, 'object or array');
+                throw new UnexpectedTypeException(sprintf(
+                    'Expected object or array, "%s" found when trying to read property "%s" of path "%s".',
+                    is_object($objectOrArray) ? get_class($objectOrArray) : gettype($objectOrArray),
```
same problem

Tobion commented on src/Symfony/Component/PropertyAccess/PropertyAccessor.php:
```diff
@@ -97,7 +97,12 @@ public function setValue(&$objectOrArray, $propertyPath, $value)
                 if (!is_object($objectOrArray) && !is_array($objectOrArray)) {
-                    throw new UnexpectedTypeException($objectOrArray, 'object or array');
+                    throw new UnexpectedTypeException(sprintf(
+                        'Expected object or array, "%s" found when trying to write property "%s" of path "%s".',
```
Please change the message to `PropertyAccessor requires a graph of objects or arrays to operate on, but it found type "%s" while trying to traverse path "%s" at property "%s"`.
Whether it's reading or writing is irrelavant at this point.

Tobion commented on src/Symfony/Component/PropertyAccess/PropertyAccessor.php:
```diff
@@ -221,7 +226,12 @@ private function &readPropertiesUntil(&$objectOrArray, PropertyPathInterface $pr
             if (!is_object($objectOrArray) && !is_array($objectOrArray)) {
-                throw new UnexpectedTypeException($objectOrArray, 'object or array');
+                throw new UnexpectedTypeException(sprintf(
+                    'Expected object or array, "%s" found when trying to read property "%s" of path "%s".',
```
Same message as above.

mpajunen committed:
[PropertyAccess] Change invalid path type exception

Use InvalidArgumentException instead of UnexpectedTypeException since
this exception is not meant to be caught.

mpajunen committed:
[PropertyAccess] Add property path to exception messages

Show the property path in the exception message when trying to access
a property of an invalid variable.

mpajunen commented:
The messages have been changed. I also realized that using a new subclass of `UnexpectedTypeException` is indeed possible. The changes are now fully BC, but had to add a new exception class and skip the parent constructor when using it.

Tobion commented on src/Symfony/Component/PropertyAccess/Exception/UnexpectedPropertyTypeException.php:
```diff
@@ -0,0 +1,34 @@
+    public function __construct($value, PropertyPathInterface $path, $index)
+    {
+        // Skip parent constructor.
+        RuntimeException::__construct(sprintf(
```
this works? looks rather hackish

mpajunen replied:
Yeah, it looks a bit hackish indeed. Seems to work well enough though.

merk replied:
It isnt a hack. https://stackoverflow.com/questions/12589289/how-to-skip-execution-of-a-parent-method-to-execute-grandparent-method

stof replied:
call the parent constructor directly rather than using this hackish solution. It will be more robust when updating code. And then, trigger the error in the parent class only when using it directly:

```php
if (__CLASS__ === get_class($this) {
    trigger_error(...);
}
```

Tobion replied:
He cannot call the parent constructor because the arguments are different.

stof replied:
ah, I missed that

Tobion commented:
If we go this route with a new subexception, then `UnexpectedTypeException` should get deprecated.

mpajunen commented:
Alright, I can add the deprecation notes in that case.

I think this approach produces the cleanest end result. The parent constructor skipping can eventually be removed when the deprecated `UnexpectedTypeException` is removed.

To avoid similar problems in the future, it might be worth considering marking exception constructors internal, or using static factory methods, or avoiding custom constructors altogether.

mpajunen commented:
Deprecation comment and error added. Change log updated as well. Should I update the UPGRADE files as well (or even add a file for 2.7)?

mpajunen committed:
[PropertyAccess] Test UnexpectedTypeException messages

mpajunen commented:
Adding a new exception class just to deprecate the old one didn't really seem that forward compatible. So I refactored the changes to use `UnexpectedTypeException` once again, but included the option to use the old arguments.

If the user code throws `UnexpectedTypeException` explicitly using the old constructor, there's a deprecation error for 2.7 and the BC layer can easily be removed for 3.0. If the constructor is not used, no user code changes will be required. A separate exception class would eventually force updating of the possible catch statements.

mpajunen committed:
[PropertyAccess] Document UnexpectedTypeException changes

mpajunen commented:
Is there something more I can do to get this merged?

(Rebased today.)

fabpot commented:
looks good to me

fabpot commented:
Thank you @mpajunen.
